iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relocatable_ptr.hpp

```diff
@@ -0,0 +1,136 @@
+///        converted to one.
+///
+/// @tparam T the native type wrapped by the relocatable_ptr, i.e. relocatable_ptr<T>
+///         has native type T and corresponds to a raw pointer of type T*.
+///
+/// @note It is advisable to use relocatable_ptr only for storage (e.g. member variables),
+///       not to pass them around as function arguments or as return value.
+///       There should be no need for this, since as pass-around type
+///       regular pointers do the job just fine and do not incur
+///       the slight runtime overhead of a relocatable_ptr.
+///       There should be no memory overhead on 64 bit systems.
+template <typename T>
+class relocatable_ptr
+{
+  public:
+    /// @brief Construct from raw pointer.
+    relocatable_ptr(T* ptr = nullptr);
+    // {
+    //     m_offset = to_offset(ptr);
+    // }
+
+    /// @brief Construct from other relocatable pointer.
+    relocatable_ptr(const relocatable_ptr& other);
+
+    /// @brief Move construct from other relocatable pointer.
+    relocatable_ptr(relocatable_ptr&& other);
+
+    /// @brief Assign from relocatable pointer rhs.
+    relocatable_ptr& operator=(const relocatable_ptr& rhs);
+
+    /// @brief Move assign from relocatable pointer rhs.
+    relocatable_ptr& operator=(relocatable_ptr&& rhs);
+
+    /// @brief Get the corresponding raw pointer.
+    /// @return corresponding raw pointer
+    T* get();
+
+    /// @brief Get the corresponding raw pointer from const relocatable_ptr
+    /// @return corresponding raw pointer
+    const T* get() const;
+
+    /// @brief Dereference a relocatable_ptr.
+    /// @return reference to the pointee
+    /// @note not available for T=void
+    template <typename S = T>
+    S& operator*();
+
+    /// @brief Dereference a const relocatable_ptr.
+    /// @return reference to the pointee
+    /// @note not available for T=void
+    template <typename S = T>
+    const S& operator*() const;
+
+    /// @brief Get the corresponding raw pointer with arrow operator syntax.
+    /// @return corresponding raw pointer
+    T* operator->();
+
+    /// @brief Get the corresponding raw pointer with arrow operator syntax
+    ///        from a const relocatable_ptr.
+    /// @return corresponding raw pointer
+    const T* operator->() const;
+
+    /// @brief Convert to the corresponding raw pointer.
+    /// @return corresponding raw pointer
+    operator T*();
+
+    /// @brief Convert to the corresponding const raw pointer.
+    /// @return corresponding const raw pointer
+    operator const T*() const;
+
+  private:
+    using offset_t = uint64_t;
+
+    // This is safe since it is equivalent to point to the relocatable pointer
+    // second byte of the relocatable pointer itself which we define to be illegal.
+    // (it is no reasonable use-case)
+    // Note that 0 is equivalent to point to the relocatable pointer itself (i.e. this).
+    static constexpr offset_t NULL_POINTER_OFFSET = 1;
```

Comment:
Why not just 0? A relocatable pointer which points to itself doesn't make sense and some CPU have an optimized instruction for comparing to 0.

Reply:
Actually only to allow pointing to itself. There should be no need for this, except for in weird data structures where `next = this` indicates some kind of end.

Reply:
Okay, in case of this
```c++
struct Foo{
relocatable_ptr<Foo>{this};
};
```
If offset 0 would encode a nullptr this wouldn't be possible anymore. So the issue is not that the pointer is referencing itself but it can have an offset of 0 in the class which contains the pointer. Could you make this more explicit in the comment.

Reply:
Okay, in case of this
```c++
struct Foo{
relocatable_ptr<Foo>{this};
};
```
If offset 0 would encode a nullptr this wouldn't be possible anymore. So the issue is not that the pointer is referencing itself but it can have an offset of 0 in the class which contains the pointer. Could you make this more explicit in the comment?